File: 362-maxProfitOfRod.py

def maxProfitUnboundedKnapsack(prices,n):  #top2bottom
    
    # @cache
    def dfs(n):
        if n<=0:
            return 0
        
        ans=0
        for i in range(n):
            ans=max(ans,dfs(n-i-1)+prices[i])
        return ans
    
    return dfs(n)


prices=[1, 3, 5, 7, 7, 7]
print(maxProfitUnboundedKnapsack(prices, 6))

File: 362-maxProfitOfRod.cpp
#include<iostream>
#include<vector>
using VEC=std::vector<int>;  
int dfs(int n,VEC& prices)  //top2bottom DP method
{
    if(n<=0)
        return 0;
    int ans=0;
    for(int i=0;i<n;++i)
    {
        ans=std::max(ans,dfs(n-i-1,prices)+prices[i]);
    }
    return ans;
}

int maxProfitBottom2Top(int n,VEC& prices) //Bottom2Top
{
    if(n<=0)
        return 0;

    VEC dp(n,0);
    dp[0]
    for(int i=0;i<n;++i)
    {
        dp[i]
    }
}


int maxProfit(int n,VEC& prices)
{
    return dfs(n,prices);
}

int main()
{
    VEC prices{1, 3, 5, 7, 7, 7};
    std::cout<<maxProfit(6,prices);
}